### ISSUE TITLE
如何得到attention map

### ISSUE THREAD

**Issue by UknowSth** [OFA-Sys/Chinese-CLIP#82]
想用中文文本来得到图片的attention map，似乎ResidualAttentionBlock中没有定义set_attn_probs函数，无法得到attn_probs的结果，如何得到图像的attention map呢。
coding能力不是很强，还请多指教，(:з」∠)

**Comment by UknowSth** [comment]
抱歉，我没有表述清楚，由于之前使用英文的CLIP模型来得到一段文本在图像中的attention map使用的是如下代码：
`def interpret(image, texts, model, device, start_layer=start_layer, start_layer_text=start_layer_text):
    batch_size = texts.shape[0]
    images = image.repeat(batch_size, 1, 1, 1)
    logits_per_image, logits_per_text = model(images, texts)
    probs = logits_per_image.softmax(dim=-1).detach().cpu().numpy()
    index = [i for i in range(batch_size)]
    one_hot = np.zeros((logits_per_image.shape[0], logits_per_image.shape[1]), dtype=np.float32)
    one_hot[torch.arange(logits_per_image.shape[0]), index] = 1
    one_hot = torch.from_numpy(one_hot).requires_grad_(True)
    one_hot = torch.sum(one_hot.cuda() * logits_per_image)
    model.zero_grad()

    image_attn_blocks = list(dict(model.visual.transformer.resblocks.named_children()).values())

    if start_layer == -1: 
      # calculate index of last layer 
      start_layer = len(image_attn_blocks) - 1
    
    num_tokens = image_attn_blocks[0].attn_probs.shape[-1]
    R = torch.eye(num_tokens, num_tokens, dtype=image_attn_blocks[0].attn_probs.dtype).to(device)
    R = R.unsqueeze(0).expand(batch_size, num_tokens, num_tokens)
    for i, blk in enumerate(image_attn_blocks):
        if i < start_layer:
            continue
        grad = torch.autograd.grad(one_hot, [blk.attn_probs], retain_graph=True)[0].detach()
        cam = blk.attn_probs.detach()
        cam = cam.reshape(-1, cam.shape[-1], cam.shape[-1])
        grad = grad.reshape(-1, grad.shape[-1], grad.shape[-1])
        cam = grad * cam
        cam = cam.reshape(batch_size, -1, cam.shape[-1], cam.shape[-1])
        cam = cam.clamp(min=0).mean(dim=1)
        R = R + torch.bmm(cam, R)
    image_relevance = R[:, 0, 1:]
    return image_relevance`
于是现在需要用中文文本来得到attention map，但似乎并不能直接通过访问attn_probs属性来得到结果。所以想要问问应该如何在前向传播以后得到attn_probs的结果。

**Comment by UknowSth** [comment]
抱歉，是我没有表述清楚，由于之前使用英文的CLIP模型来得到一段文本在图像中的attention map使用的是如下代码：
```
def interpret(image, texts, model, device, start_layer=start_layer, start_layer_text=start_layer_text):
batch_size = texts.shape[0]
images = image.repeat(batch_size, 1, 1, 1)
logits_per_image, logits_per_text = model(images, texts)
probs = logits_per_image.softmax(dim=-1).detach().cpu().numpy()
index = [i for i in range(batch_size)]
one_hot = np.zeros((logits_per_image.shape[0], logits_per_image.shape[1]), dtype=np.float32)
one_hot[torch.arange(logits_per_image.shape[0]), index] = 1
one_hot = torch.from_numpy(one_hot).requires_grad_(True)
one_hot = torch.sum(one_hot.cuda() * logits_per_image)
model.zero_grad()

image_attn_blocks = list(dict(model.visual.transformer.resblocks.named_children()).values())

if start_layer == -1: 
  # calculate index of last layer 
  start_layer = len(image_attn_blocks) - 1

num_tokens = image_attn_blocks[0].attn_probs.shape[-1]
R = torch.eye(num_tokens, num_tokens, dtype=image_attn_blocks[0].attn_probs.dtype).to(device)
R = R.unsqueeze(0).expand(batch_size, num_tokens, num_tokens)
for i, blk in enumerate(image_attn_blocks):
    if i < start_layer:
        continue
    grad = torch.autograd.grad(one_hot, [blk.attn_probs], retain_graph=True)[0].detach()
    cam = blk.attn_probs.detach()
    cam = cam.reshape(-1, cam.shape[-1], cam.shape[-1])
    grad = grad.reshape(-1, grad.shape[-1], grad.shape[-1])
    cam = grad * cam
    cam = cam.reshape(batch_size, -1, cam.shape[-1], cam.shape[-1])
    cam = cam.clamp(min=0).mean(dim=1)
    R = R + torch.bmm(cam, R)
image_relevance = R[:, 0, 1:]
return image_relevance
```

于是现在需要用中文文本来得到attention map，但似乎并不能直接通过访问attn_probs属性来得到结果。所以想要问问应该如何在前向传播以后得到attn_probs的结果。

**Comment by DtYXs** [comment]
> 

您好，我的理解是您代码里的`attn_probs`就是`self.attn(x, x, x, need_weights=False, attn_mask=self.attn_mask)[1]`
您可以试试看修改`cn_clip/clip/model.py`中`ResidualAttentionBlock`类的相关代码，添加上`attn_probs`这个属性，供您参考（=========标记部分为修改部分）：
```python
class ResidualAttentionBlock(nn.Module):
    def __init__(self, d_model: int, n_head: int, attn_mask: torch.Tensor = None, use_flash_attention: bool = False):
        super().__init__()

        self.attn = nn.MultiheadAttention(d_model, n_head) if not use_flash_attention else FlashMHA(d_model, n_head)
        self.ln_1 = LayerNorm(d_model)
        self.mlp = nn.Sequential(OrderedDict([
            ("c_fc", nn.Linear(d_model, d_model * 4)),
            ("gelu", QuickGELU()),
            ("c_proj", nn.Linear(d_model * 4, d_model))
        ]))
        self.ln_2 = LayerNorm(d_model)
        self.attn_mask = attn_mask
        self.use_flash_attention = use_flash_attention
        # ===================
        self.attn_probs = None
        # ===================

    def attention(self, x: torch.Tensor):
        self.attn_mask = self.attn_mask.to(dtype=x.dtype, device=x.device) if self.attn_mask is not None else None
        if self.use_flash_attention:
            # Batch first is needed for FlashAttention. See https://github.com/HazyResearch/flash-attention/issues/84 for more information.
            return self.attn(x.transpose(1, 0))[0].transpose(1, 0)
        else:
            # ===================
            self.attn_probs = self.attn(x, x, x, need_weights=False, attn_mask=self.attn_mask)[1]
            # ===================
            return self.attn(x, x, x, need_weights=False, attn_mask=self.attn_mask)[0]
```

**Comment by UknowSth** [comment]
非常感谢您的解答，解决了。谢谢

**Comment by WangshilongChina** [comment]
> 非常感谢您的解答，解决了。谢谢

我在运行可视化的时候报错， grad = torch.autograd.grad(one_hot, [blk.attn_probs], retain_graph=True,allow_unused=True)[0].detach() ，One of the differentiated Tensors appears to not have been used in the graph. Set allow_unused=True if this is the desired behavior. 请问您是怎么解决的？
